### PR TITLE
Added support to anyopen for taking filelike objects; remove broken `.zip` support

### PR DESCRIPTION
**Pull request by dotsdl**
Use in this way requires specifying the compression, if any, that the stream features in order to decompress it.

### CODE REVIEW

**Review by dwhswenson — COMMENTED** [review]
Thanks! It looks like this will largely work for OpenFE's needs. I've tested it out with our in-progress storage, and we'll just have to be sure we handle encodings on our end when that is needed.

A couple minor comments/suggestions from looking over the code.

File: src/alchemlyb/parsing/util.py [hunk 1]
```diff
@@ -39,12 +48,27 @@ def anyopen(filename, mode='r'):
...
                   '.zip': zipfile.ZipFile}
 
-    ext = os.path.splitext(filename)[1]
+    # compression selections available
+    compressions = {'bz2': bz2_open,
+                    'gz': gzip_open,
+                    'zip': zipfile.ZipFile}
+
+    # if `datafile` is a stream
+    if hasattr(datafile, 'read'):
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply src/alchemlyb/parsing/util.py]: So far `anyopen` hasn't been used with writing in mind in `alchemlyb`, but there's no reason it couldn't be I suppose. Added explicit checks for certain modes, and tests for roundtripping through a file and a stream.

File: src/alchemlyb/parsing/util.py [hunk 2]
```diff
@@ -14,19 +16,26 @@ def gzip_open(filename, mode):
...
-    filename : str
-        Path to file to use.
+    datafile : PathLike | IO
+        Path to file to use, or an open IO stream. If an IO stream, use
+        `compression` to specify the type of compression used, if any.
     mode : str
         Mode for stream; usually 'r' or 'w'.
+    compression : str
+        Use to specify compression.
+        Must be one of 'bz2', 'gz', 'zip'.
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply src/alchemlyb/parsing/util.py]: Ah, good catch. I've changed the behavior such that `compression`, if specified, *always* forces the compression assumed by `anyopen`.

File: src/alchemlyb/parsing/util.py [hunk 3]
```diff
@@ -39,12 +48,27 @@ def anyopen(filename, mode='r'):
                   '.gz': gzip_open,
                   '.zip': zipfile.ZipFile}
 
-    ext = os.path.splitext(filename)[1]
+    # compression selections available
+    compressions = {'bz2': bz2_open,
+                    'gz': gzip_open,
+                    'zip': zipfile.ZipFile}
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply src/alchemlyb/parsing/util.py]: Adjusted compression dict to make things more explicit, in which we specify either `bzip2` or `gzip` for compression.

File: src/alchemlyb/tests/parsing/test_util.py [hunk 4]
```diff
@@ -12,3 +14,42 @@ def test_gzip():
         for filename in dataset['data'][leg]:
             with anyopen(filename, 'r') as f:
                 assert type(f.readline()) is str
+
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply src/alchemlyb/tests/parsing/test_util.py:18]: Added explicit test forcing compression despite wrong extension.

**Comment by dotsdl** [comment]
Thank you @dwhswenson for the review! I've implemented changes in response to your feedback.

File: src/alchemlyb/parsing/util.py [hunk 5]
```diff
@@ -10,41 +12,71 @@ def bz2_open(filename, mode):
...
-        Path to file to use.
+    datafile : PathLike | IO
+        Path to file to use, or an open IO stream. If an IO stream, use
+        `compression` to specify the type of compression used, if any.
     mode : str
         Mode for stream; usually 'r' or 'w'.
+    compression : str
+        Use to specify compression. Must be one of 'bzip2', 'gzip'.
+        Overrides use of extension for determining compression if `datafile` is
+        a file.
```
Comment (orbeckst) [inline src/alchemlyb/parsing/util.py:41]: add versionadded

File: src/alchemlyb/parsing/util.py [hunk 6]
```diff
@@ -10,41 +12,71 @@ def bz2_open(filename, mode):
...
     -------
     stream : stream
-        Open stream for reading.
+        Open stream for reading or writing, depending on mode.
 
     """
     # opener for each type of file
     extensions = {'.bz2': bz2_open,
-                  '.gz': gzip_open,
-                  '.zip': zipfile.ZipFile}
```
Comment (orbeckst) [inline src/alchemlyb/parsing/util.py:40]: Why did you remove zip?

Comment (orbeckst) [reply src/alchemlyb/parsing/util.py:40]: ... and remove `import zipfile`

**Review by orbeckst — CHANGES_REQUESTED** [review]
Thanks for the upgrade. Small comments on docs. Coverage indicates that there's still quite a bit uncovered code, please increase testing. Please add an entry to CHANGES.

**Comment by orbeckst** [comment]
@dotsdl any chance you can come back to the PR and address remaining comments? Would be nice to get it in.

**Comment by dotsdl** [comment]
@orbeckst yes, apologies, was on vacation through middle of last week and looping back to this shortly. Thank you for your review!

File: src/alchemlyb/parsing/util.py (hunk 6 above)
Comment (dotsdl) [reply src/alchemlyb/parsing/util.py:40]: Zip doesn't work like `gzip` or `bz2` because it's an archive format, not merely a compressor. Reading a "zip-compressed" file requires specifying the file's name in the archive as well. 

The functionality we intended to have here likely never worked (mid-twenties me is to blame :sweat_smile:), so I'm removing it. We don't have test coverage of it, either, nor any `zip`-archived files in `alchemtest`.

File: src/alchemlyb/parsing/util.py [hunk 7]
```diff
@@ -10,41 +12,71 @@ def bz2_open(filename, mode):
...
+        elif compression in compressions:
+            compressor = compressions[compression]
+            return compressor(datafile, mode=mode)
+        else:
+            raise ValueError("`datafile` is a stream, but specified `compression` '{compression}' is not supported")
 
-    if ext in extensions:
-       opener= extensions[ext]
+    # otherwise, treat as a file
+    # allow compression to override any extension on the file
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply src/alchemlyb/parsing/util.py:82]: Done!

File: src/alchemlyb/parsing/util.py (hunk 6 above)
Comment (dotsdl) [reply src/alchemlyb/parsing/util.py:40]: Addressed the above items!

File: src/alchemlyb/parsing/util.py [hunk 8]
```diff
@@ -10,41 +12,71 @@ def bz2_open(filename, mode):
...
+    compression : str
+        Use to specify compression. Must be one of 'bzip2', 'gzip'.
+        Overrides use of extension for determining compression if `datafile` is
+        a file.
 
     Returns
     -------
     stream : stream
-        Open stream for reading.
+        Open stream for reading or writing, depending on mode.
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply src/alchemlyb/parsing/util.py:55]: Done!

File: src/alchemlyb/parsing/util.py (hunk 5 above)
Comment (dotsdl) [reply src/alchemlyb/parsing/util.py:48]: Done!

File: src/alchemlyb/parsing/util.py (hunk 6 above)
Comment (orbeckst) [reply src/alchemlyb/parsing/util.py:40]: Thanks for the explanation!

File: src/alchemlyb/parsing/util.py [hunk 9]
```diff
@@ -2,49 +2,93 @@
...
+    if ((hasattr(datafile, 'read') and any((i in mode for i in ('r',)))) or 
+       (hasattr(datafile, 'write') and any((i in mode for i in ('w', 'a', 'x'))))):
+        # if no compression specified, just pass the stream through
+        if compression is None:
+            return datafile
+        elif compression in compressions:
+            compressor = compressions[compression]
+            return compressor(datafile, mode=mode)
+        else:
+            raise ValueError("`datafile` is a stream, but specified `compression` '{compression}' is not supported")
```
Comment (orbeckst) [inline src/alchemlyb/parsing/util.py:79]: This is the only line that needs testing.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Please add a test for the case where the wrong `compression` is set on a stream. Otherwise I am very happy with everything — thank you @dotsdl !

File: src/alchemlyb/parsing/util.py (hunk 9 above)
Comment (dotsdl) [reply src/alchemlyb/parsing/util.py:79]: Done!

**Comment by orbeckst** [comment]
I'll get #200 fixed first and then re-run CI here.

**Review by orbeckst — APPROVED** [review]
Thank you for addressing my remaining comment. All good! Thanks!

**Comment by dotsdl** [comment]
@orbeckst danke! :pray:

**Comment by orbeckst** [comment]
(The checks had passed before I added everything to CHANGES and I because I need to get some supper, I took a short cut...)
